### DISCUSSION TITLE
Can SeleniumVBA be used in VBScript?

### DISCUSSION THREAD

**Discussion by GCuser99** [GCuser99/SeleniumVBA#71] [General]
The [SeleniumVBA DLL solution](https://github.com/GCuser99/SeleniumVBA/tree/main/dist) can be used in your VBScripts, with a few caveats. 

To demonstrate, let's try to replicate one of the test subroutines in VBA using a VBScript. In the SeleniumVBA test routines there is a test_download_resource subroutine that looks like this in VBA:
```vba
Sub test_download_resource()
    'this test uses the DownloadResource method of the WebElement class to download the src to an img element
    Dim driver As WebDriver
    Dim element As WebElement

    Set driver = New WebDriver

    driver.StartChrome
    driver.OpenBrowser
    
    driver.NavigateTo "https://github.com/GCuser99/SeleniumVBA/wiki"
    driver.Wait 1000

    Set element = driver.FindElement(By.cssSelector, "img[alt='SeleniumVBA'")
    
    'if a folder path is specified for fileOrFolderPath, then the saved file inherits the name of the source
    element.DownloadResource srcAttribute:="src", fileOrFolderPath:=".\"
    
    driver.CloseBrowser
    driver.Shutdown
End Sub
```
The VBA subroutine above demonstrates the DownloadResource method by downloading the SeleniumVBA logo file from GitHub repository.

To replicate this using VBScript, create a text file called "test_download_resource.vbs" and open it with a text editor such as notepad. Copy/paste the following into the file:

```vba
'this test uses the DownloadResource method of the WebElement class to download the src to an img element

'Use late-bound CreateObject to instantiate a new SeleniumVBA object
Set driver = CreateObject("SeleniumVBA.WebDriver")

driver.StartChrome
driver.OpenBrowser
    
driver.NavigateTo "https://github.com/GCuser99/SeleniumVBA/wiki"
driver.Wait 1000

'VBScript does not support Enums so cannot use the By enumerator.
'Set element = driver.FindElement(By.cssSelector, "img[alt='SeleniumVBA'")

'There are several work-arounds for above but the easiest fix is to replace with its alternative method:
Set element = driver.FindElementByCssSelector("img[alt='SeleniumVBA'")

'VBScript does not recognize the Named Argument syntax (":=") of passing argument values
'element.DownloadResource srcAttribute:="src", fileOrFolderPath:=".\"

'So must pass argument values the other way:
'element.DownloadResource "src", ".\"

'BUT, use of relative path like ".\" will end up pointing to the DLL library folder, not the folder containing the script...

'So if that is not the desired behavior, then best practice is to use absolute paths like below:
element.DownloadResource "src", "%USERPROFILE%\Documents\SeleniumVBA\scripts\"
    
driver.CloseBrowser
driver.Shutdown
```
After pasting the above into the file, modify the folder path "%USERPROFILE%\Documents\SeleniumVBA\scripts\" to a valid absolute path of your choosing to save the logo file to. Then save and close the file (make sure you gave the file a "vbs" extension!).

From the Windows File Explorer, double click on the saved VBScript file. You will see the twinBasic banner for 5 seconds, then the browser navigation to the web page. After the script has executed, you should find the logo.png file in whichever folder that you specified above. It's that easy.

For more examples of using SeleniumVBA in VBScript, see the [vbscript folder](https://github.com/GCuser99/SeleniumVBA/tree/main/dev/twinBasic/vbscript) in this repository.

In summary, the use of SeleniumVBA in VBScript can provide a light-weight alternative to using it in MS Excel and Access when those applications are not central to the scripting objective.

Have an example to share, a question, or suggestion? Feel free to contribute to this discussion!

**Comment by 6DiegoDiego9** [comment]
Thanks for the example!
I'm going to use VBScript just as launchers of XLSM or XLAM clones to obtain multitasking, but for very simple scripts this could be a light weight alternative, I agree.

About CreateObject("SeleniumVBA.WebDriver"), does it require the registration of the DLL, right?

**Reply by GCuser99** [reply]
> About CreateObject("SeleniumVBA.WebDriver"), does it require the registration of the DLL, right?

Yes, it does. But it's on my list to look into what we could do with a standard (non-ActiveX) DLL...

**Comment by GCuser99** [comment]
Apparently VBScript does not support DLL API declarations, so even though the SeleniumVBA DLL can be used without registration using API declarations to instantiate the class objects (late bound of course), I could not find a direct way to this with VBScript. Bummer...

**Comment by 6DiegoDiego9** [comment]
Not my area of expertise and I didn't try that, anyway this is what the new Bing answers:

![image](https://user-images.githubusercontent.com/109476941/219480746-25f53490-56da-44f7-8603-a0954827994f.png)

Source:  https://stackoverflow.com/a/9846377
